### ISSUE TITLE
How to install marlin-binary-protocol for FirmwareUpdater plugin?

### ISSUE THREAD

**Issue by Xeevis** [jneilliii/OctoPrint-WindowsInstaller#21]
Hello, I'm trying setup FirmwareUpdater, but it's complaining about missing `marlin-binary-protocol` package.

![image](https://github.com/jneilliii/OctoPrint-WindowsInstaller/assets/5835044/00ba5b0a-455d-4260-87a6-b9155cf6b8c9)

I've followed [these instructions](https://github.com/OctoPrint/OctoPrint-FirmwareUpdater/blob/devel/doc/marlinbft.md#marlin-binary-protocol-package), but I must be doing something wrong.

I've installed Python 3.10 from Microsoft Store and tried running commands like
```
CMD> C:\Users\Xeevis\AppData\Local\Programs\OctoPrint\WPy64-31050\python-3.10.5.amd64\Lib\venv\scripts\nt\activate.bat
```

But I have no idea what I'm doing and it just keeps on piling errors. Any idea what steps should I take on Windows?

**Comment by jneilliii** [comment]
The windows installer does it slightly different than the venv approach mentioned in those instructions. You would have to run the following command to do pip installs into the OctoPrint environment. 

```
c:\OctoPrint\WPy64-31050\scripts\python.bat -m pip install https://github.com/p3p/pyheatshrink/releases/download/0.3.3/pyheatshrink-pip.zip
```

however, I just tried that from my machine and it gives compile errors.

I suspect the issue is related to python 3.10 as the heatshrink module does say that python 3 support is experimental.

**Comment by jneilliii** [comment]
actually after doing the next step of installing marlin-binary-file-transfer, it is using heatshrink2 now.

![image](https://github.com/jneilliii/OctoPrint-WindowsInstaller/assets/5249455/a8cad679-8cc1-457b-9ead-d2e8a0b2bf94)

so really you just need to run this I think.

```
C:\Users\Xeevis\AppData\Local\Programs\OctoPrint\WPy64-31050\scripts\python.bat -m pip install marlin-binary-protocol
```

**Comment by Xeevis** [comment]
Thanks a ton this did the trick! I've made PowerShell one-liner to simplify the process:
```
cd $env:LOCALAPPDATA\Programs\OctoPrint\WPy64-31050\scripts\ && Invoke-WebRequest -Uri https://alfter.us/wp-content/uploads/2022/08/pyheatshrink-92ffa4e.tar.xz -OutFile $env:temp\pyheatshrink-92ffa4e.tar.xz && ./python.bat -m pip install $env:temp\pyheatshrink-92ffa4e.tar.xz && ./python.bat -m pip install marlin-binary-protocol
```

![image](https://github.com/jneilliii/OctoPrint-WindowsInstaller/assets/5835044/b00b9aab-fb4d-4e59-b36a-9d6302e870f6)

**Comment by rozhkovets** [comment]
When I try install from https://alfter.us/wp-content/uploads/2022/08/pyheatshrink-92ffa4e.tar.xz  I get an error that the file stdlib.h was not found

what am I missing?

```
C:\OctoPrint\WPy64-31050\scripts>pip install https://alfter.us/wp-content/uploads/2022/08/pyheatshrink-92ffa4e.tar.xz
Collecting https://alfter.us/wp-content/uploads/2022/08/pyheatshrink-92ffa4e.tar.xz
  Downloading https://alfter.us/wp-content/uploads/2022/08/pyheatshrink-92ffa4e.tar.xz (1.9 MB)
     ---------------------------------------- 1.9/1.9 MB 713.6 kB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: Heatshrink
  Building wheel for Heatshrink (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [19 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-310
      creating build\lib.win-amd64-cpython-310\heatshrink
      copying heatshrink\streams.py -> build\lib.win-amd64-cpython-310\heatshrink
      copying heatshrink\__init__.py -> build\lib.win-amd64-cpython-310\heatshrink
      running build_ext
      building 'heatshrink.core' extension
      creating build\temp.win-amd64-cpython-310
      creating build\temp.win-amd64-cpython-310\Release
      creating build\temp.win-amd64-cpython-310\Release\heatshrink
      creating build\temp.win-amd64-cpython-310\Release\heatshrink\_heatshrink
      "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.38.33130\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -I. -I./heatshrink/_heatshrink -IC:\OctoPrint\WPy64-31050\python-3.10.5.amd64\include -IC:\OctoPrint\WPy64-31050\python-3.10.5.amd64\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.38.33130\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" /Tcheatshrink/_heatshrink/heatshrink_decoder.c /Fobuild\temp.win-amd64-cpython-310\Release\heatshrink/_heatshrink/heatshrink_decoder.obj -std=c99
      cl: Є®¬ ­¤­ п бва®Є  warning D9002: Їа®ЇгбЄ ­ҐЁ§ўҐбв­®Ј® Ї а ¬Ґва  "-std=c99"
      heatshrink_decoder.c
      heatshrink/_heatshrink/heatshrink_decoder.c(1): fatal error C1083: ЌҐ г¤ Ґвбп ®вЄалвм д ©« ўЄ«озҐ­ЁҐ: stdlib.h: No such file or directory,
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.38.33130\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for Heatshrink
  Running setup.py clean for Heatshrink
Failed to build Heatshrink
Installing collected packages: Heatshrink
  Running setup.py install for Heatshrink ... error
  error: subprocess-exited-with-error

  × Running setup.py install for Heatshrink did not run successfully.
  │ exit code: 1
  ╰─> [21 lines of output]
      running install
      C:\OctoPrint\WPy64-31050\python-3.10.5.amd64\lib\site-packages\setuptools\command\install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-310
      creating build\lib.win-amd64-cpython-310\heatshrink
      copying heatshrink\streams.py -> build\lib.win-amd64-cpython-310\heatshrink
      copying heatshrink\__init__.py -> build\lib.win-amd64-cpython-310\heatshrink
      running build_ext
      building 'heatshrink.core' extension
      creating build\temp.win-amd64-cpython-310
      creating build\temp.win-amd64-cpython-310\Release
      creating build\temp.win-amd64-cpython-310\Release\heatshrink
      creating build\temp.win-amd64-cpython-310\Release\heatshrink\_heatshrink
      "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.38.33130\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -I. -I./heatshrink/_heatshrink -IC:\OctoPrint\WPy64-31050\python-3.10.5.amd64\include -IC:\OctoPrint\WPy64-31050\python-3.10.5.amd64\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.38.33130\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" /Tcheatshrink/_heatshrink/heatshrink_decoder.c /Fobuild\temp.win-amd64-cpython-310\Release\heatshrink/_heatshrink/heatshrink_decoder.obj -std=c99
      cl: Є®¬ ­¤­ п бва®Є  warning D9002: Їа®ЇгбЄ ­ҐЁ§ўҐбв­®Ј® Ї а ¬Ґва  "-std=c99"
      heatshrink_decoder.c
      heatshrink/_heatshrink/heatshrink_decoder.c(1): fatal error C1083: ЌҐ г¤ Ґвбп ®вЄалвм д ©« ўЄ«озҐ­ЁҐ: stdlib.h: No such file or directory,
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.38.33130\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> Heatshrink

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.

[notice] A new release of pip available: 22.1.2 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip
```
